### ISSUE TITLE
Proposal: QLoRA + DeepSpeed memory-efficiency example

### ISSUE THREAD

**Issue by hallleedustin-alt** [yiqiao-yin/deepspeed-course#32]
I’m interested in contributing a QLoRA example that compares standard LoRA with 4-bit QLoRA using the same base model, dataset, and training steps.

My goal would be to focus primarily on infrastructure and memory efficiency rather than model performance. I’d like to measure peak GPU memory, trainable parameters, training time, and whether loss decreases during a short training run.

I’m considering Qwen3-4B as the default model so the example can be tested relatively inexpensively on a single RunPod GPU.

I noticed that the repository contains LoRA and some quantization material, but I did not find a standalone QLoRA + DeepSpeed example comparing the memory profiles.

Would this make more sense as a new numbered topic because of the different memory profile, or as an extension of an existing LLM fine-tuning topic?

**Comment by yiqiao-yin** [comment]
Yes please — this is a good proposal, and your read of the gap is accurate. I checked: 4-bit appears in exactly one place, `03_llms/03_ocr`, where it is a flag on a *vision-language* example running at ZeRO-2, and `bitsandbytes` is locked in that one folder only. There is no standalone QLoRA-vs-LoRA comparison, and the framing you want — infrastructure and memory rather than model quality — is the right one for this course.

## Your question: new numbered topic

**A new numbered topic**, not an extension. Two reasons:

- A topic here is a **folder**, not a page. It needs its own `ds_config*.json`, its own `pyproject.toml` + committed `uv.lock` (yours would pull `bitsandbytes`, which only `03_ocr` currently locks), its own launcher and its own test. Bolting a second memory profile onto an existing folder would break the property that a reader can open one directory and run it.
- What you are actually building is a **comparison harness**, not a fine-tune. That is its own thing, and it deserves a name that says so.

`03_llms/12_qlora` is the next free slot. `uv run scripts/new_example.py 03_llms/12_qlora --title "QLoRA vs LoRA" --vram 24` writes the skeleton with the contract already satisfied, and `./tests/run_all.sh` is green before you write any of your own code.

## One thing that will decide your whole design, so please check it first

**Do not run the quantized arm under ZeRO-3.** In transformers 5.16.1 — the version every lab here pins — `modeling_utils.py` guards **twelve** separate ZeRO-3 parameter paths with `and not is_quantized`:

```python
if is_deepspeed_zero3_enabled() and not is_quantized:
```

So when the model is quantized, `zero.Init`'s parameter partitioning is **skipped**. The run does not fail. It trains, the loss falls, and the base weights are quietly replicated in full on every rank — which means a memory comparison done that way is measuring something other than what its labels claim. This is the exact failure mode the repo keeps running into: it exits 0 and the number is meaningless.

**Use ZeRO-2 for both arms.** ZeRO-2 shards optimizer state and gradients but leaves parameters alone, which is fine for a frozen quantized base and keeps the comparison apples-to-apples. If you do want to say something about ZeRO-3, make it a stated finding with evidence rather than a config choice.

I would genuinely like to see that measured rather than taken on my word — if you find it behaves differently, that is a more interesting result than the one we are expecting.

## What would make the measurement trustworthy

Your four metrics are the right ones. A few traps that have each cost someone here a day:

- **Budget and measure per GPU, never in aggregate.** `per_GPU = weights/N + overhead that does not shard`. Activations, gather buffers and fragmentation are paid in full by every rank. An aggregate check passed 2 × 48 GB for a 55.6 GB model that then OOMed at the first step.
- **`p.numel()` returns 0 under ZeRO-3** once `zero.Init` has partitioned — use `ds_numel`. And in the *4-bit* arm, `bitsandbytes` packs two parameters per `uint8`, so a naive `numel()` on `Params4bit` does not mean what it looks like. Whatever you report, report how you counted it.
- **Reset the peak between arms** (`torch.cuda.reset_peak_memory_stats()`), or arm two inherits arm one's high-water mark and the comparison collapses.
- **Assert the two arms actually differ.** A sweep in `04_video_text/06_qwen3vl` once printed a flat 14.2 GB at 4, 8, 16 and 24 frames and exited 0 — it had measured one input four times, and it looked exactly like a memory ceiling. If your harness can be misconfigured into running the same arm twice, make it **raise** rather than print a tidy table.
- **A short run must not be reported as a failure.** If `--max-steps 20` is offered, do not print "Poor — train longer" under a "Finished Successfully" banner. Say the run was capped and what a real one looks like.

For the test, the property worth asserting is the *relationship*, not an absolute: QLoRA peak < LoRA peak, both losses decrease, trainable-parameter counts match expectation. Absolute GB numbers vary by card and will make the suite flaky. And please run the check against a deliberately broken version first — a test you have not watched fail is not a test yet.

## Qwen3-4B is a good default

4.02 B parameters, so bf16 weights are 8.04 GB and 4-bit is ~2.01 GB plus quantization state. Both arms fit a single 24 GB card with room for the comparison to be visible, which is exactly what you want. (For contrast, full fine-tuning would need ~32 GB of Adam state alone — worth a line in the README, since it is the reason anyone reaches for LoRA at all.)

## The contract

`CONTRIBUTING.md` is the spec and is written to be followed end to end; §3 and §4 are the parts that block a merge. In short, your example needs to work for three readers:

1. **No GPU** — `require_gpu()` called *before* torch/deepspeed are imported, a message saying what they can still do, `ALLOW_CPU=1` honoured, and a **non-zero exit**. Test it with `CUDA_VISIBLE_DEVICES="" uv run python train_qlora.py; echo "exit=$?"` — a guard that prints beautifully and exits 0 tells a calling script the run succeeded.
2. **CoreWeave** — `run_deepspeed.sh` with `#SBATCH` headers and a cheap `--max-steps` dry-run path.
3. **RunPod** — an `EXAMPLES` entry in `runpod/runpod_ctl.py`, and a README documenting `run <ex> --dry-run --collect --wait --terminate --yes` plus confirming with `pods`.

Plus the six files (`train_*.py`, `ds_config*.json`, `run_deepspeed.sh`, `README.md`, `pyproject.toml`, **committed `uv.lock`**), a `clawdeck.yaml` entry, a test registered in **both** `tests/run_all.sh` and `.github/workflows/tests.yml`, and a docs page listed in `sidebars.js`.

`uv run scripts/check_contract.py 03_llms/12_qlora` checks most of these for you. Everything is `uv` — never bare `pip` — and the lock must be committed, because `cd 03_llms/12_qlora && uv sync` has to work from a fresh clone.

I have just updated `CONTRIBUTING.md` with several things that were only in maintainer notes, including two RunPod defaults that will bite you: `--wait-seconds` defaults to 1800 s, and `--disk` is the *container* disk while `--volume` (default 40 GB) is what mounts at `/workspace` where `HF_HOME` points. Worth a skim before you start.

## If you cannot test on hardware

Say so in the PR and register the table entry anyway — someone can verify on a GPU. **Do not invent expected output.** A fabricated number costs the next reader a day debugging a setup that was fine.

Happy to review a draft PR early rather than at the end, if it would save you rework. Thanks for offering this one — it fills a gap I would not have prioritised and should have.
